fix(clusterpools): cluster-pool-ipv4-cidr/cluster-pool-ipv6-cidr

cndoit18 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

```release-note
`cluster-pool-ipv4-cidr` and `cluster-pool-ipv6-cidr` options now accept string slices and not just string
```

cndoit18 commented on operator/flags.go:
```diff
@@ -171,7 +171,7 @@ func init() {
 	option.BindEnv(option.EnableIPv4Name)
 
-	flags.String(operatorOption.ClusterPoolIPv4CIDR, "",
+	flags.StringSlice(operatorOption.ClusterPoolIPv4CIDR, []string{""},
```
see 	https://github.com/cilium/cilium/blob/f6dcb4ffe4e5933fbdecb9f4d27937ef292c42a6/operator/option/config.go#L292

errordeveloper replied:
Ack!

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -707,11 +707,11 @@ data:
 {{- end }}
 {{- if .Values.ipv6.enabled }}
-  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | quote }}
+  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | join " " |quote }}
```
I don't think using space works as a separator for pflag.StringSlice.

```suggestion
  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | join "," |quote }}
```

cndoit18 replied:
https://github.com/cilium/cilium/blob/f3597588e0bbc29045a669f759cc4aa99dc34953/pkg/option/config.go#L2330

I checked this place and if "," is used, it is judged to be a whole string and not a list of strings.
Because the parameters are read from `ReadDirConfig`

gandro replied:
Ah, you seem to be correct correct! Thanks for checking!

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -925,10 +930,14 @@ ipam:
   operator:
     # -- IPv4 CIDR range to delegate to individual nodes for IPAM.
     clusterPoolIPv4PodCIDR: "10.0.0.0/8"
```
```suggestion
    # -- Deprecated in favor of ipam.operator.clusterPoolIPv4PodCIDRList.
    # IPv4 CIDR range to delegate to individual nodes for IPAM.
    clusterPoolIPv4PodCIDR: "10.0.0.0/8"
```

errordeveloper commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -106,13 +106,13 @@ affinity:
                 - linux
   podAntiAffinity:
     requiredDuringSchedulingIgnoredDuringExecution:
-    - labelSelector:
```
This and other indentation changes to `value.yaml` seem unnecessary. If you would like to cleanup indentation, please put that into a separate commit, or better - another PR :)

cndoit18 replied:
Sorry, these are caused by my own formatting tools

errordeveloper commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -127,7 +127,8 @@ extraEnv: {}
 # -- Additional agent hostPath mounts.
-extraHostPathMounts: []
+extraHostPathMounts:
+  []
```
This and similar changes below don't match the style in which this file is currently written...

joestringer replied:
Indeed, was there a particular tool you used to format them this way? I don't think that the previous developers working on this file were aware of any such tool.

cndoit18 replied:
> Indeed, was there a particular tool you used to format them this way? I don't think that the previous developers working on this file were aware of any such tool.

https://github.com/redhat-developer/yaml-language-server

Yes. my vscode editor will automatically format it

errordeveloper commented on operator/flags.go:
```diff
@@ -178,7 +178,7 @@ func init() {
 	option.BindEnv(option.EnableIPv4Name)
 
-	flags.String(operatorOption.ClusterPoolIPv4CIDR, "",
+	flags.StringSlice(operatorOption.ClusterPoolIPv4CIDR, []string{},
```
The change of field type doesn't seem to be included in this commit, did you forget to run `git add`?

gandro replied:
Not necessary. The code actually already reads it as a string slice. See https://github.com/cilium/cilium/pull/17780#discussion_r742735839 and https://github.com/cilium/cilium/blob/858fca7e2cfa78d3b5faa8d28086c2d8581fde8c/operator/option/config.go#L376

errordeveloper replied:
I see, thanks for clarifying.

errordeveloper reviewed (changes requested):
Please see my comments

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -925,10 +930,14 @@ ipam:
     # -- IPv4 CIDR mask size to delegate to individual nodes for IPAM.
     clusterPoolIPv4MaskSize: 24
     # -- IPv6 CIDR range to delegate to individual nodes for IPAM.
     clusterPoolIPv6PodCIDR: "fd00::/104"
```
```suggestion
    # -- Deprecated in favor of ipam.operator.clusterPoolIPv6PodCIDRList.
    # IPv6 CIDR range to delegate to individual nodes for IPAM.
    clusterPoolIPv6PodCIDR: "fd00::/104"
```

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -707,11 +707,11 @@ data:
 {{- if eq $ipam "cluster-pool" }}
 {{- if .Values.ipv4.enabled }}
-  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | quote }}
+  cluster-pool-ipv4-cidr: {{ append .Values.ipam.operator.clusterPoolIPv4PodCIDRList .Values.ipam.operator.clusterPoolIPv4PodCIDR | join " " | quote }}
```
I think it's better to use an `if-else` statement here:

The use of `append` kind of makes it hard to use non-default values for `.ipam.operator.clusterPoolIPv4PodCIDR`.

For example, if I want to have `[192.168.0.0/16, 172.16.0.0/12]` as my pod CIDRs, naively I would specify the following value: `ipam.operator.clusterPoolIPv4PodCIDRList=[192.168.0.0/16, 172.16.0.0/12]`

But that would be wrong, since this `append` will pull in the default value of  `10.0.0.0/8` too, so I accidentally would end up with `[192.168.0.0/16, 172.16.0.0/12 10.0.0.0/8]`

I would have to set `ipam.operator.clusterPoolIPv4PodCIDR=172.16.0.0/12` **and** `ipam.operator.clusterPoolIPv4PodCIDRList=[192.168.0.0/16]` to achieve my desired result, which I think is a bit confusing.

Therefore, my suggestion would be to use the list if specified, and fall back on the deprecated value if not:
```suggestion
  {{- if .Values.ipam.operator.clusterPoolIPv4PodCIDRList }}
  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDRList | join " " | quote }}
  {{- else }}
  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | quote }}
  {{- end }}
```

gandro reviewed (commented):
Some inline feedback around the deprecation of the old value.

Thanks a lot!

cndoit18 committed:
fix(cluster-pools): fix parameter

`cluster-pool-ipv4-cidr` and `cluster-pool-ipv6-cidr` should be string slices and not string

Signed-off-by: cndoit18 <cndoit18@outlook.com>

gandro reviewed (approved):
Looks good to me! Awesome, thank you for taking care of this!

joestringer reviewed (changes requested):
Thanks for the PR. At a glance this seems like it should be safe for upgrades from existing clusterpool configuration, although I have not tested it out.

There's a few references to the old name for this field in the docs, could you fix them up?

```
$ git grep clusterPoolIPv4
Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4PodCIDR=<IPv4CIDR>``
Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4MaskSize=<IPv4MaskSize>``
Documentation/gettingstarted/k8s-install-openshift-okd.rst:reflect those changes. Namely ``clusterNetwork`` should match ``nativeRoutingCIDR``, ``clusterPoolIPv4PodCIDR`` and ``clusterPoolIPv4MaskSize``.
...
```

Also, should this be added into the `Documentation/operations/upgrade.rst` upgrade guide to inform users that they should migrate their configurations to the new settings?

joestringer reviewed (approved):
LGTM, thanks.
